numpyro/infer/enum_messenger.py

```diff
@@ -0,0 +1,557 @@
+        return tuple(batch_shape)
+
+    def process_message(self, msg):  # copied almost verbatim from plate
+        if msg['type'] not in ('sample',):
+            return
+        cond_indep_stack = msg['cond_indep_stack']
+        frame = CondIndepStackFrame(self.name, self.dim, self.subsample_size)
+        cond_indep_stack.append(frame)
+        expected_shape = self._get_batch_shape(cond_indep_stack)
+        dist_batch_shape = msg['fn'].batch_shape if msg['type'] == 'sample' else ()
+        overlap_idx = max(len(expected_shape) - len(dist_batch_shape), 0)
+        trailing_shape = expected_shape[overlap_idx:]
+        # e.g. distribution with batch shape (1, 5) cannot be broadcast to (5, 5)
+        broadcast_shape = lax.broadcast_shapes(trailing_shape, dist_batch_shape)
+        if broadcast_shape != dist_batch_shape:
+            raise ValueError('Distribution batch shape = {} cannot be broadcast up to {}. '
+                             'Consider using unbatched distributions.'
+                             .format(dist_batch_shape, broadcast_shape))
+        batch_shape = expected_shape[:overlap_idx]
+        if 'sample_shape' in msg['kwargs']:
+            batch_shape = lax.broadcast_shapes(msg['kwargs']['sample_shape'], batch_shape)
+        msg['kwargs']['sample_shape'] = batch_shape
+        if self.size != self.subsample_size:
+            scale = 1. if msg['scale'] is None else msg['scale']
+            msg['scale'] = scale * self.size / self.subsample_size
+
+
+class enum(BaseEnumMessenger):
+    """
+    This version of EnumMessenger uses to_data to allocate a fresh enumeration dim
+    for each discrete sample site.
+    """
+    def process_message(self, msg):
+
+        if msg["type"] != "sample" or \
+                msg["done"] or msg["is_observed"] or msg["infer"].get("expand", False) or \
+                msg["infer"].get("enumerate") != "parallel":
+            return
+
+        if msg["infer"].get("num_samples", None) is not None:
+            raise NotImplementedError("TODO implement multiple sampling")
+
+        if msg["infer"].get("expand", False):
+            raise NotImplementedError("expand=True not implemented")
+
+        raw_value = msg["fn"].enumerate_support(expand=False).squeeze()
+        size = raw_value.numel()
+        funsor_value = funsor.Tensor(
+            raw_value,
+            OrderedDict([(msg["name"], funsor.bint(size))]),
+            size
+        )
+
+        msg["value"] = to_data(funsor_value)
+        msg["done"] = True
+
+
+class trace(OrigTraceMessenger):
+    """
+    This version of TraceMessenger records information necessary to do packing after execution.
+    Each sample site is annotated with a "dim_to_name" dictionary,
+    which can be passed directly to funsor.to_funsor.
+    """
+    def postprocess_message(self, msg):
+        if msg["type"] == "sample":
+            msg["infer"]["dim_to_name"] = NamedMessenger._get_dim_to_name(msg["fn"].batch_shape)
```

Comment:
This metadata should be sufficient to pack the `log_prob` tensors for sum-product contraction with `funsor.sum_product` in a modified `potential_energy` implementation.